libs/backend/src/cast_vote_record_report_import.ts

```diff
@@ -0,0 +1,221 @@
+  | {
+      type: 'invalid-directory';
+      message: string;
+    }
+  | { type: 'missing-report'; message: string }
+  | { type: 'missing-layouts'; message: string };
+
+/**
+ * Result of validating the cast vote record report directory structure.
+ * Either a list of relative ballot image file paths or an error.
+ */
+export type ValidateCastVoteRecordReportDirectoryStructureResult = Result<
+  string[],
+  CastVoteRecordReportDirectoryStructureValidationError
+>;
+
+function invalidDirectoryError(
+  error: ListDirectoryError
+): ValidateCastVoteRecordReportDirectoryStructureResult {
+  return err({
+    type: 'invalid-directory',
+    message: error.message,
+  });
+}
+
+/**
+ * Validates that the directory fits the structure expected of a cast vote
+ * record report directory. If successful, returns a list of the contained
+ * ballot images with paths relative to the ballot images root. If not
+ * successful, returns {@link CastVoteRecordReportDirectoryStructureValidationError}.
+ * This does *not* validate that all ballot images referenced in the report are
+ * included in the directory, only that all ballot images have accompanying layouts.
+ */
+export async function validateCastVoteRecordReportDirectoryStructure(
+  directoryAbsolutePath: string
+): Promise<ValidateCastVoteRecordReportDirectoryStructureResult> {
+  const listDirectoryRootResult = await listDirectory(directoryAbsolutePath);
+  if (listDirectoryRootResult.isErr()) {
+    return invalidDirectoryError(listDirectoryRootResult.err());
+  }
+
+  const rootFileEntries = listDirectoryRootResult.ok();
+
+  const reportFileEntry = rootFileEntries.find(
+    (fileEntry) => fileEntry.name === CAST_VOTE_RECORD_REPORT_FILENAME
+  );
+  if (!reportFileEntry) {
+    return err({
+      type: 'missing-report',
+      message: `Expected a cast vote record report with filename ${CAST_VOTE_RECORD_REPORT_FILENAME} but none was found.`,
+    });
+  }
+
+  const ballotImageDirectoryEntry = rootFileEntries.find(
+    (fileEntry) => fileEntry.name === CVR_BALLOT_IMAGES_SUBDIRECTORY
+  );
+
+  if (!ballotImageDirectoryEntry) {
+    return ok([]);
+  }
+
+  const ballotLayoutsDirectoryEntry = rootFileEntries.find(
+    (fileEntry) => fileEntry.name === CVR_BALLOT_LAYOUTS_SUBDIRECTORY
+  );
+
+  if (!ballotLayoutsDirectoryEntry) {
+    return err({
+      type: 'missing-layouts',
+      message:
+        'Expected ballot images to be accompanied by ballot layouts, but none were found.',
+    });
+  }
+
+  const listBallotImageContentsResult = await listDirectoryRecursive(
+    join(directoryAbsolutePath, CVR_BALLOT_IMAGES_SUBDIRECTORY)
+  );
+  if (listBallotImageContentsResult.isErr()) {
+    return invalidDirectoryError(listBallotImageContentsResult.err());
+  }
+  const ballotImageFiles = listBallotImageContentsResult
+    .ok()
+    .filter(
+      (fileEntry) =>
+        fileEntry.type !== FileSystemEntryType.Directory &&
+        fileEntry.name.endsWith('.jpg')
```

Comment:
I just know someone in the future is going to spend an inordinate amount of time figuring out why it's not working and they'll eventually realize it's because the images are `.jpeg` instead of `.jpg`.